synapse/replication/tcp/resource.py

```diff
@@ -117,6 +117,16 @@ def on_notifier_poke(self):
                 stream.discard_updates_and_advance()
             return
 
+        # We check up front to see if anything has actually changed, as we get
+        # poked because of changes that happened on other instances.
+        if all(
+            stream.last_token == stream.current_token(self._instance_name)
+            for stream in self.streams
+        ):
+            return
```

Comment:
I wondered if there was a clever way we could filter out the streams that have not changed and only operate on the ones that have changed, but I think that would require a much large refactoring.

For reference, this seems to be uplifting the check in the loop (https://github.com/matrix-org/synapse/blob/9627934c851b906002b17c97096c7f828e30b915/synapse/replication/tcp/resource.py#L158-L161) to handle a case of skipping all entries.

Reply:
Yeah, I thought about doing something more clever then decided life was too short :slightly_smiling_face: